frontend/model/notifications/nativeNotification.js

```diff
@@ -4,50 +4,112 @@ import sbp from '@sbp/sbp'
-  if (force || Notification.permission === 'default') {
-    try {
-      sbp('state/vuex/commit', 'setNotificationEnabled', await Notification.requestPermission() === 'granted')
-    } catch (e) {
-      console.error('requestNotificationPermission:', e.message)
-      return null
+const fallbackChangeListener = () => {
+  if (!Notification.permission) return
+  let oldValue = Notification.permission
+  handler([oldValue])
+  setInterval(() => {
+    const newValue = Notification.permission
+    if (oldValue !== newValue) {
+      oldValue = newValue
+      handler([oldValue])
     }
-  }
+  }, 100)
+}
+
+export const setupNativeNotificationsListeners = () => {
+  if (typeof navigator !== 'object' || typeof Notification !== 'function' || typeof PushManager !== 'function' || typeof ServiceWorker !== 'function' || typeof navigator.serviceWorker !== 'object') return
 
-  if (Notification.permission === 'granted') {
-    await sbp('service-worker/setup-push-subscription').catch((e) => {
-      console.error('Error setting up service worker', e)
+  if (
+    typeof navigator.permissions === 'object' &&
+    // $FlowFixMe[method-unbinding]
+    typeof navigator.permissions.query === 'function'
+  ) {
+    Promise.all([
+      navigator.permissions.query({ name: 'notifications' }),
+      navigator.permissions.query({ name: 'push', userVisibleOnly: true })
+    ]).then(
+      (statuses) => {
+        const states = statuses.map(status => status.state)
+        handler(states)
+        statuses[0].addEventListener('change', () => {
+          states[0] = statuses[0].state
+          handler(states)
+        }, false)
+        statuses[1].addEventListener('change', () => {
+          states[1] = statuses[1].state
+          handler(states)
+        }, false)
+      }
+    ).catch((e) => {
+      console.error('Error querying notifications permission', e)
+      fallbackChangeListener()
     })
+  } else {
+    fallbackChangeListener()
+  }
+}
+
+export async function requestNotificationPermission (): Promise<null | string> {
+  if (typeof Notification !== 'function') {
+    return null
   }
 
-  return Notification.permission
+  try {
+    const result = await Notification.requestPermission()
+    sbp('state/vuex/commit', 'setNotificationEnabled', result === 'granted')
+    return result
+  } catch (e) {
+    console.error('requestNotificationPermission:', e.message)
+    return null
+  }
 }
 
 export function makeNotification ({ title, body, icon, path }: {
   title: string, body: string, icon?: string, path?: string
 }): void {
-  if (Notification?.permission === 'granted' && sbp('state/vuex/settings').notificationEnabled) {
+  if (typeof Notification !== 'function') return
+  // If not running on a SW
+  if (typeof window === 'object') {
     try {
+      // $FlowFixMe[incompatible-type]
+      if (navigator.vendor === 'Apple Computer, Inc.') {
+        throw new Error('Safari requires a service worker for the notification to be displayed')
+      }
```

Comment:
Claude:

```js
// Current problematic check:
if (navigator.vendor === 'Apple Computer, Inc.') {

// Should be more comprehensive, as this could include other WebKit browsers:
if (navigator.vendor.includes('Apple') && /^((?!chrome|android).)*safari/i.test(navigator.userAgent)) {
```

Reply:
Claude is wrong in this case, because Chrome doesn't have this issue and also can't be used for notifications in iOS. The test was specifically made to target Safari, in a rare instance of using browser sniffing as this can't be feature-detected.

Reply:
And "as this could include other WebKit browsers:"

Not sure what this means exactly, but I expect all WebKit browsers to have this issue, but also it's hard to test because there are no WebKit browsers that people actually use other than Safari. There's GNOME Web, but it doesn't support notifications, and then there is MiniBrowser (which is meant for testing, not serious browsing) which also doesn't support notifications.

Reply:
I think Claude's check isn't about checking Chrome though, but rather other browsers that might be WebKit based

Reply:
Claude is checking if the vendor contains 'Apple' (my check does this but more exactly) and that the UA string doesn't contain `chrome` nor `android` (case insensitively). This check seems arbitrary because I'm not aware of any browsers that exist and that people use that (1) are WebKit-based, (2) include 'Apple' in `navigator.vendor` (granted, all WebKit browsers probably do), (3) importantly, support push notifications, and (4) for some reason don't have this bug and (5) have a different bug that makes the workaround fail.

In practice, _and I've tried_, outside of the Apple ecosystem, the _only_ WebKit-based browser is GNOME Web, which doesn't seem to support push notifications. `new Notification()` _does_ work in GNOME Web, although it doesn't include `chrome` nor `android` in the user agent. All iOS browsers are WebKit-based, however this check is irrelevant for them because `Notification` isn't at all supported in iOS outside of PWAs, and Chrome for iOS doesn't include `chrome` at all, it includes `CriOS`.

However, the workaround using the SW registration should work in all browsers that support notifications and service workers. I could remove the check for Safari and unconditionally use the fallback, and the only reason I haven't is because I suspect that `new Notification` is slightly more performant than going through the SW.

So, what I'm saying is that this is an instance of Claude making stuff up because the additional check doesn't match any browser in existence, nor any browser likely to be created in the future. I guess theoretically it'd match _ancient_ versions of Chrome, from the time Chrome was WebKit-based, which we already don't support.